File: domain/entity.ts
/**
 * domain/entity.ts
 *
 * Entity is the pure domain model for a typed node in the knowledge graph —
 * a person, organization, device, account, location, or asset — distinct
 * from Identity, which aggregates the identifiers that resolve to a subject.
 * An Entity may optionally point back to the Identity it belongs to.
 *
 * Mirrors contracts/domain/entity.schema.json field-for-field.
 *
 * Domain Purity: standard runtime and sibling domain modules only.
 */

import {
  Status,
  freezeRefs,
  freezeTags,
  newId,
  now,
  type Id,
  type Metadata,
  type Provenance,
  type Reference,
  type Timestamp,
  EMPTY_METADATA,
} from "./common.js";
import { ClassificationLevel } from "./classification_level.js";
import { ConfidenceScore } from "./confidence_score.js";

export enum EntityType {
  Person = "person",
  Organization = "organization",
  Device = "device",
  Account = "account",
  Location = "location",
  Asset = "asset",
  Document = "document",
  Other = "other",
}

export interface EntityInit {
  readonly entityType: EntityType;
  readonly name?: string;
  readonly identityRef?: Reference;
  readonly attributes?: Metadata;
  readonly status?: Status;
  readonly classification?: ClassificationLevel;
  readonly relationshipRefs?: ReadonlyArray<Reference>;
  readonly evidenceRefs?: ReadonlyArray<Reference>;
  readonly sourceRefs?: ReadonlyArray<Reference>;
  readonly confidence?: ConfidenceScore;
  readonly provenance: Provenance;
  readonly tags?: ReadonlyArray<string>;
  readonly metadata?: Metadata;
}

/**
 * Entity is an immutable node in the knowledge graph. Relationship
 * attachments are tracked by reference only; the Entity itself never holds
 * a live pointer to a Relationship instance, preserving layer boundaries.
 */
export class Entity {
  readonly id: Id;
  readonly version: string;
  readonly createdAt: Timestamp;
  readonly updatedAt: Timestamp;
  readonly status: Status;
  readonly classification: ClassificationLevel;
  readonly entityType: EntityType;
  readonly name?: string;
  readonly identityRef?: Reference;
  readonly attributes: Metadata;
  readonly relationshipRefs: ReadonlyArray<Reference>;
  readonly evidenceRefs: ReadonlyArray<Reference>;
  readonly sourceRefs: ReadonlyArray<Reference>;
  readonly confidence?: ConfidenceScore;
  readonly provenance: Provenance;
  readonly tags: ReadonlyArray<string>;
  readonly metadata: Metadata;

  private constructor(id: Id, version: string, createdAt: Timestamp, updatedAt: Timestamp, init: EntityInit) {
    this.id = id;
    this.version = version;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.status = init.status ?? Status.Active;
    this.classification = init.classification ?? ClassificationLevel.Internal;
    this.entityType = init.entityType;
    this.name = init.name;
    this.identityRef = init.identityRef;
    this.attributes = init.attributes ?? EMPTY_METADATA;
    this.relationshipRefs = freezeRefs(init.relationshipRefs);
    this.evidenceRefs = freezeRefs(init.evidenceRefs);
    this.sourceRefs = freezeRefs(init.sourceRefs);
    this.confidence = init.confidence;
    this.provenance = init.provenance;
    this.tags = freezeTags(init.tags);
    this.metadata = init.metadata ?? EMPTY_METADATA;
    Object.freeze(this);
  }

  static create(init: EntityInit): Entity {
    const timestamp = now();
    return new Entity(newId(), "1.0.0", timestamp, timestamp, init);
  }

  /** Returns a new Entity with the given relationship reference attached. */
  withRelationshipRef(relationshipRef: Reference): Entity {
    if (this.relationshipRefs.some((ref) => ref.refId === relationshipRef.refId)) {
      return this;
    }
    return new Entity(this.id, this.version, this.createdAt, now(), {
      entityType: this.entityType,
      name: this.name,
      identityRef: this.identityRef,
      attributes: this.attributes,
      status: this.status,
      classification: this.classification,
      relationshipRefs: [...this.relationshipRefs, relationshipRef],
      evidenceRefs: this.evidenceRefs,
      sourceRefs: this.sourceRefs,
      confidence: this.confidence,
      provenance: this.provenance,
      tags: this.tags,
      metadata: this.metadata,
    });
  }
}

File: domain/entity.py
"""
domain/entity.py

Entity domain model - a distinguishable thing of interest in an investigation.
Mirrors contracts/domain/entity.schema.json field-for-field.
"""
from dataclasses import dataclass, field
from enum import Enum
from typing import FrozenSet, Sequence

from .classification_level import ClassificationLevel, ClassificationLevelValue
from .confidence_score import ConfidenceScore
from .common import (
    DomainInvariantError,
    EMPTY_METADATA,
    Metadata,
    Provenance,
    Reference,
    Status,
    new_id,
    utc_now,
)


class EntityType(str, Enum):
    """Types of entities that can be distinguished in investigations."""
    PERSON = "person"
    ORGANIZATION = "organization"
    LOCATION = "location"
    DEVICE = "device"
    ACCOUNT = "account"
    SERVICE = "service"
    CONTENT = "content"
    EVENT = "event"
    CONCEPT = "concept"
    OTHER = "other"


@dataclass(frozen=True)
class EntityAttributes:
    """Optional typed attributes for an entity."""
    attributes: dict = field(default_factory=dict)


@dataclass(frozen=True)
class Entity:
    """
    Entity is a distinguishable thing of interest in an investigation.
    
    Entities are linked to Identities and can be referenced by Evidence,
    Findings, and other Entities.
    """
    id: str
    version: str
    created_at: str
    updated_at: str
    entity_type: EntityType
    name: str
    status: Status
    classification: ClassificationLevel
    provenance: Provenance
    identity_ref: Reference | None = None
    evidence_refs: FrozenSet[Reference] = field(default_factory=frozenset)
    finding_refs: FrozenSet[Reference] = field(default_factory=frozenset)
    relationship_refs: FrozenSet[Reference] = field(default_factory=frozenset)
    confidence: ConfidenceScore | None = None
    attributes: EntityAttributes = field(default_factory=EntityAttributes)
    tags: FrozenSet[str] = field(default_factory=frozenset)
    metadata: Metadata = field(default_factory=lambda: dict(EMPTY_METADATA))

    @classmethod
    def create(
        cls,
        entity_type: EntityType,
        name: str,
        provenance: Provenance,
        identity_ref: Reference | None = None,
        status: Status = Status.ACTIVE,
        classification: ClassificationLevelValue = ClassificationLevelValue.INTERNAL,
        evidence_refs: Sequence[Reference] | None = None,
        finding_refs: Sequence[Reference] | None = None,
        relationship_refs: Sequence[Reference] | None = None,
        confidence: ConfidenceScore | None = None,
        attributes: dict | None = None,
        tags: Sequence[str] | None = None,
        metadata: Metadata | None = None,
    ) -> "Entity":
        """Create a new Entity."""
        if not name or not name.strip():
            raise DomainInvariantError("Entity.name must not be empty.")
        
        timestamp = utc_now()
        return cls(
            id=new_id(),
            version="1.0.0",
            created_at=timestamp,
            updated_at=timestamp,
            entity_type=entity_type,
            name=name,
            status=status,
            classification=ClassificationLevel(classification),
            identity_ref=identity_ref,
            evidence_refs=frozenset(evidence_refs) if evidence_refs else frozenset(),
            finding_refs=frozenset(finding_refs) if finding_refs else frozenset(),
            relationship_refs=frozenset(relationship_refs) if relationship_refs else frozenset(),
            confidence=confidence,
            provenance=provenance,
            attributes=EntityAttributes(attributes) if attributes else EntityAttributes(),
            tags=frozenset(tags) if tags else frozenset(),
            metadata=dict(metadata) if metadata else dict(EMPTY_METADATA),
        )

    def with_evidence_ref(self, ref: Reference) -> "Entity":
        """Return a new Entity with the given evidence reference added."""
        if ref in self.evidence_refs:
            return self
        return Entity(
            id=self.id,
            version=self.version,
            created_at=self.created_at,
            updated_at=utc_now(),
            entity_type=self.entity_type,
            name=self.name,
            status=self.status,
            classification=self.classification,
            identity_ref=self.identity_ref,
            evidence_refs=self.evidence_refs | {ref},
            finding_refs=self.finding_refs,
            relationship_refs=self.relationship_refs,
            confidence=self.confidence,
            provenance=self.provenance,
            attributes=self.attributes,
            tags=self.tags,
            metadata=self.metadata,
        )


__all__ = [
    "Entity",
    "EntityAttributes",
    "EntityType",
]
